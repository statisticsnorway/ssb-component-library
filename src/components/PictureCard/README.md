# PictureCard

> Clickable image box with title

### Usage

#### HTML

```html
<a class="ssb-picture-card" href=" ">
  <div class="image-background"><img src="" alt="" /></div>
  <span class="il-type">Type</span>
  <span class="il-title">Title</span>
  {feather.arrowRightCircle 32px}
</a>
```

#### React

```jsx harmony
<PictureCard imageSrc={url} altText="alt text" onClick={function} type="Type" title="Tittel" />
```

```jsx harmony
<PictureCard imageSrc={url} altText='alt text' link='./' type='Type' title='Tittel' />
```

Available props:

| Name      | Type            | Description                                  |
| --------- | --------------- | -------------------------------------------- |
| className | string          | Optional container class                     |
| imageSrc  | required (url)  | URL of the image to be used                  |
| altText   | required string | alt text for screen readers                  |
| link      | string          | Link to be navigated to                      |
| onClick   | func            | Function callback to be used instead of link |
| title     | string          | Title text                                   |
| type      | string          | Type text                                    |
