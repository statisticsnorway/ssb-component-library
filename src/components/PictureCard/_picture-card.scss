.ssb-picture-card {
  overflow: hidden;
  position: relative;
  display: flex;
  align-items: flex-end;
  width: 280px;
  height: 400px;
  color: $ssb-white !important;
  cursor: pointer;

  .image-background {
    img {
      top: 0;
      left: 0;
      position: absolute;
      transform: scale(1);
      transition:
        transform 0.25s ease,
        -webkit-transform 0.25s ease;
    }
  }

  &:hover {
    img {
      transform: scale(1.1);
    }
  }

  &:focus {
    @include focus-ring;
  }

  .overlay {
    bottom: 0;
    position: absolute;
    padding: 80px 20px 64px;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    width: 280px;
    background-image: linear-gradient(0deg, rgb(39 66 71 / 75%), rgb(39 66 71 / 50%), transparent);

    .il-type {
      @include roboto;
      font-size: 16px;
      line-height: 1.25;
      margin-bottom: 12px;
    }

    .il-title {
      @include roboto-condenced;
      font-size: 28px;
      font-weight: bold;
      line-height: 1.43;
      margin-bottom: 12px;
    }
  }

  &.horizontal {
    height: 212px;
    padding: 40px 20px;
    width: 380px;

    .image-background {
      height: 212px;
      width: 380px;
    }

    .overlay {
      width: 380px;
      padding: 40px 20px;
      left: 0;
    }
  }
}
