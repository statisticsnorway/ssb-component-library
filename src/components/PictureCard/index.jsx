import React, { useRef, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { ArrowRight, ArrowRightCircle } from 'react-feather';

export function useHover() {
	const [value, setValue] = useState(false);

	const hoverRef = useRef(null);

	useEffect(
		// eslint-disable-next-line consistent-return
		() => {
			const handleMouseOver = () => setValue(true);
			const handleMouseOut = () => setValue(false);
			const element = hoverRef && hoverRef.current;

			if (element) {
				element.addEventListener('mouseover', handleMouseOver);
				element.addEventListener('mouseout', handleMouseOut);
				return () => {
					element.removeEventListener('mouseover', handleMouseOver);
					element.removeEventListener('mouseout', handleMouseOut);
				};
			}
		},
		[hoverRef],
	);

	return [hoverRef, value];
}

const PictureCard = ({ className, imageSrc, altText, link, onClick, orientation, title, type }) => {
	const [hoverRef, hovered] = useHover();
	return (
		<a
			className={`ssb-picture-card ${orientation} ${className || ''}`}
			href={link}
			onClick={onClick}
			ref={hoverRef}
		>
			<div className="image-background">
				<img src={imageSrc} alt={altText} />
			</div>
			<div className="overlay">
				<span className="il-type">{type}</span>
				<span className="il-title">{title}</span>
				{hovered
					? <ArrowRightCircle className="arrow-icon" size={32} />
					: <ArrowRight className="arrow-icon" size={32} />}
			</div>
		</a>
	);
};

PictureCard.defaultProps = {
	onClick: () => {
	},
	orientation: 'vertical',
};

PictureCard.propTypes = {
	className: PropTypes.string,
	imageSrc: PropTypes.string.isRequired,
	altText: PropTypes.string.isRequired,
	link: PropTypes.string,
	onClick: PropTypes.func,
	orientation: PropTypes.oneOf(['horizontal', 'vertical']),
	title: PropTypes.string,
	type: PropTypes.string,
};

export default PictureCard;
