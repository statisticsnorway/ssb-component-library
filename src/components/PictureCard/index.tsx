import React, { forwardRef, useRef, useState, useEffect, type MutableRefObject } from 'react'
import { ArrowRight, ArrowRightCircle } from 'react-feather'

interface PictureCardProps {
  className?: string
  imageSrc: string
  altText: string
  link?: string
  onClick: () => void
  orientation: 'horizontal' | 'vertical'
  title?: string
  type?: string
}

export function useHover(): [MutableRefObject<HTMLAnchorElement | null>, boolean] {
  const [value, setValue] = useState(false)

  const hoverRef = useRef<HTMLAnchorElement>(null)
  useEffect(() => {
    const handleMouseOver = () => setValue(true)
    const handleMouseOut = () => setValue(false)
    const element = hoverRef.current

    if (element) {
      element.addEventListener('mouseover', handleMouseOver)
      element.addEventListener('mouseout', handleMouseOut)
      return () => {
        element.removeEventListener('mouseover', handleMouseOver)
        element.removeEventListener('mouseout', handleMouseOut)
      }
    }

    // Return a no-op function if element is not defined
    return () => {}
  }, [hoverRef])

  return [hoverRef, value]
}

const PictureCard = forwardRef<HTMLAnchorElement, PictureCardProps>(
  ({ className, imageSrc, altText, link, onClick = () => {}, orientation = 'vertical', title, type }, ref) => {
    const [hoverRef, hovered] = useHover()
    const setRefs = (element: HTMLAnchorElement | null) => {
      // Using ref for multiple purposes, so need to set it manually
      const mutableRef = ref as React.MutableRefObject<HTMLAnchorElement | null>
      if (typeof ref === 'function') {
        ref(element)
      } else if (mutableRef) {
        mutableRef.current = element
      }
      hoverRef.current = element
    }

    return (
      <a className={`ssb-picture-card ${orientation} ${className || ''}`} href={link} onClick={onClick} ref={setRefs}>
        <div className='image-background'>
          <img src={imageSrc} alt={altText} />
        </div>
        <div className='overlay'>
          <span className='il-type'>{type}</span>
          <span className='il-title'>{title}</span>
          {hovered ? (
            <ArrowRightCircle className='arrow-icon' size={32} />
          ) : (
            <ArrowRight className='arrow-icon' size={32} />
          )}
        </div>
      </a>
    )
  }
)

export default PictureCard

/*
ref={element => {
	// Using ref for multiple purposes, so need to set it manually
	if (typeof ref === 'function') ref(element);
	// eslint-disable-next-line no-param-reassign
	else if (ref) { ref.current = element; }
	hoverRef.current = element;
}}

The ref is set in this way so it can be both used by a parent component with useRef and the internal hoverRef hook.
The 'if (typeof ref === 'function')'  is used so a parent can use the two commonly used syntaxes for setting refs

Both of these will now work:
 <PictureCard ref={(element) => (cards.current[index] = element)}
 <PictureCard ref={myRef}
The first example is the most used way to set refs on components in .map functions.
*/
