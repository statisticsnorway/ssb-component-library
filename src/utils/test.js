import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
