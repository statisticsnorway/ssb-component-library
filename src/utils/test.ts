import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
