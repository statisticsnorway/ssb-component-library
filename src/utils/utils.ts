// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
