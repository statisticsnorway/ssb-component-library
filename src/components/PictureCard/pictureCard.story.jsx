import React from 'react'
import PictureCard from './index'
// import cardImage from './cards_bilde.jpg';

const sup = () => {
  console.log('Image link clicked')
}

export default {
  title: 'Picture Card',
}

export const Default = () => (
  <PictureCard imageSrc='/cards_bilde.jpg' altText='img alt' onClick={sup} type='Type' title='Tittel' />
)

export const Horizontal = () => (
  <PictureCard
    orientation='horizontal'
    imageSrc='/cards_bilde.jpg'
    altText='img alt'
    onClick={sup}
    type='Type'
    title='Tittel'
  />
)
