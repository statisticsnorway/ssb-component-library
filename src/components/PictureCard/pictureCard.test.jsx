import React from 'react'
import { render } from '../../utils/test'
import PictureCard from './index'
import cardImage from './cards_bilde.jpg'

describe('ImageLink component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(
      <PictureCard imageSrc={cardImage} onClick={() => {}} type='Type' altText='img alt' title='Tittel' />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('Sets default orientation className', () => {
    const { asFragment } = render(
      <PictureCard imageSrc={cardImage} altText='alt img' onClick={() => {}} type='Type' title='Tittel' />
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('Switches className on orientation change', () => {
    const { asFragment } = render(
      <PictureCard
        orientation='horizontal'
        imageSrc={cardImage}
        altText='alt img'
        onClick={() => {}}
        type='Type'
        title='Tittel'
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
