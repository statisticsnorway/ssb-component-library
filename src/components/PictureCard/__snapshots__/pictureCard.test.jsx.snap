// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageLink component Matches the snapshot 1`] = `
<DocumentFragment>
  <a
    class="ssb-picture-card vertical "
  >
    <div
      class="image-background"
    >
      <img
        alt="img alt"
        src=""
      />
    </div>
    <div
      class="overlay"
    >
      <span
        class="il-type"
      >
        Type
      </span>
      <span
        class="il-title"
      >
        Tittel
      </span>
      <svg
        class="arrow-icon"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`ImageLink component Sets default orientation className 1`] = `
<DocumentFragment>
  <a
    class="ssb-picture-card vertical "
  >
    <div
      class="image-background"
    >
      <img
        alt="alt img"
        src=""
      />
    </div>
    <div
      class="overlay"
    >
      <span
        class="il-type"
      >
        Type
      </span>
      <span
        class="il-title"
      >
        Tittel
      </span>
      <svg
        class="arrow-icon"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </div>
  </a>
</DocumentFragment>
`;

exports[`ImageLink component Switches className on orientation change 1`] = `
<DocumentFragment>
  <a
    class="ssb-picture-card horizontal "
  >
    <div
      class="image-background"
    >
      <img
        alt="alt img"
        src=""
      />
    </div>
    <div
      class="overlay"
    >
      <span
        class="il-type"
      >
        Type
      </span>
      <span
        class="il-title"
      >
        Tittel
      </span>
      <svg
        class="arrow-icon"
        fill="none"
        height="32"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="32"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </div>
  </a>
</DocumentFragment>
`;
